### ISSUE TITLE
Install-Module -Name Az hangs for long without progress and makes sounds

### ISSUE THREAD

**Issue by arphox** [Azure/azure-powershell#17210]
### Description

Hi,

I was actually creating a bug report in this repository because the install module command took so much time before actually reporting progerss, but then it started showing symptoms of doing something so I decided to raise this problem instead.

Steps I took:
1. Start PowerShell 7.2.1 as administrator in Windows Terminal
2. Run `Set-ExecutionPolicy -ExecutionPolicy RemoteSigned -Scope CurrentUser`
3. Run `Install-Module -Name Az -Scope CurrentUser -Repository PSGallery -Force`

**At this point, the installation hangs for minutes without any output.**

Literally, I just started it and it does nothing for several minutes:
![image](https://user-images.githubusercontent.com/17320094/155023728-ce4ed3bb-7f42-40cc-829b-dcd59fd31b30.png)

This is one issue.

Another issue is that during installation, it makes annoying sounds (I think those are Windows notification sounds) while installing, I think when it installs a part of it.

My PC is modern, my internet speed is good, I am not behind any firewall or proxy or anything that would justify this horribly slow install process.

For example, as I am typing this, the `Installing dependent package 'Az.Maintenance'` part is running for several minutes, so I think I will just leave it like this for the night.

**Comment by arphox** [comment]
Also, the installation tutorial lacks a part of "how to tell if you have it installed already". e.g. I don't know that.

**Comment by scerazy** [comment]
It does indeed take "forever" to install!

**Comment by ghubz** [comment]
Same here. Linux install, it's been stuck here for I don't know how long..

`PS /root> Install-Module -Name Az -Scope CurrentUser -Repository PSGallery -Force`

**Comment by dingmeng-xue** [comment]
Install-Module doesn't belong to Azure PowerShell. In my experience, the problem usually is due to the network or performance of PSGallery. You can measure the performance via 
```
```

**Comment by arphox** [comment]
@dingmeng-xue 
the mentioned line of code does not work.

**See:**
<details>
  <summary>Spoiler</summary>
  
```
PS C:\Users\ozsvartk> (1,2,3) | ForEach-Object { Measure-Command {Find-Module Az-repository PSGallery}}
Find-Module: A positional parameter cannot be found that accepts argument 'PSGallery'.

Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 2
Milliseconds      : 930
Ticks             : 29307940
TotalDays         : 3,39212268518519E-05
TotalHours        : 0,000814109444444444
TotalMinutes      : 0,0488465666666667
TotalSeconds      : 2,930794
TotalMilliseconds : 2930,794

Find-Module: A positional parameter cannot be found that accepts argument 'PSGallery'.
Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 0
Milliseconds      : 2
Ticks             : 22463
TotalDays         : 2,59988425925926E-08
TotalHours        : 6,23972222222222E-07
TotalMinutes      : 3,74383333333333E-05
TotalSeconds      : 0,0022463
TotalMilliseconds : 2,2463

Find-Module: A positional parameter cannot be found that accepts argument 'PSGallery'.
Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 0
Milliseconds      : 3
Ticks             : 36683
TotalDays         : 4,24571759259259E-08
TotalHours        : 1,01897222222222E-06
TotalMinutes      : 6,11383333333333E-05
TotalSeconds      : 0,0036683
TotalMilliseconds : 3,6683


PS C:\Users\ozsvartk>
```

</details>

**Nevertheless, my problem is that the installation process lacks transparence: it is not clear what it is doing.**
If this is not on this software component, please redirect this issue to whoever is responsible for this.

**Comment by dingmeng-xue** [comment]
Got it. 
```
```

**Comment by toutougabi** [comment]
WE have the same issue on 7.2.1 nothing seems to be happening, CPU is mostly at 0 but goes to 3-4% once in a while

**Comment by jiannystein** [comment]
Same issue on 7.2.2 
![image](https://user-images.githubusercontent.com/41028418/160959377-eb5b3d6a-ceba-49fa-a2fc-176b489f4dfd.png)

**Comment by CannonHax** [comment]
It really is running slow.  I executed the "Install-Module -Name Az -Scope CurrentUser -Repository PSGallery -Force" command over 5 minutes ago, and I'm not seeing anything yet.

**Comment by CannonHax** [comment]
Ok, finally seeing some action

![image](https://user-images.githubusercontent.com/81432601/161089392-f755a765-1312-467a-828a-b1edb02e9dfe.png)

**Comment by remcoros** [comment]
"No recent activity" is probably because we are still waiting on the installation to finish.

Anything we can do to provide more detail?

**Comment by CannonHax** [comment]
In my case, it took a great deal of patience, but eventually the installation started and completed.

**Comment by remcoros** [comment]
Seems a performance issue on PSGallery. Each "searching repository....." call below takes about 10 - 30 seconds.

And it's definitely not on my end, have been pulling and pushing docker images over my 1Gbit fiber connection all day today.

`
VERBOSE: Using the provider 'PowerShellGet' for searching packages.
VERBOSE: The -Repository parameter was not specified.  PowerShellGet will use all of the registered repositories.
VERBOSE: Getting the provider object for the PackageManagement Provider 'NuGet'.
VERBOSE: The specified Location is 'https://www.powershellgallery.com/api/v2' and PackageManagementProvider is 'NuGet'.
VERBOSE: Searching repository 'https://www.powershellgallery.com/api/v2/FindPackagesById()?id='Az'' for ''.
VERBOSE: Total package yield:'1' for the specified package 'Az'.
VERBOSE: Performing the operation "Install-Module" on target "Version '7.4.0' of module 'Az'".
VERBOSE: The installation scope is specified to be 'CurrentUser'.
VERBOSE: The specified module will be installed in 'C:\Users\Remco\Documents\PowerShell\Modules'.
VERBOSE: The specified Location is 'NuGet' and PackageManagementProvider is 'NuGet'.
VERBOSE: Downloading module 'Az' with version '7.4.0' from the repository 'https://www.powershellgallery.com/api/v2'.
VERBOSE: Searching repository 'https://www.powershellgallery.com/api/v2/FindPackagesById()?id='Az'' for ''.
VERBOSE: Searching repository 'https://www.powershellgallery.com/api/v2/FindPackagesById()?id='Az.Accounts'' for ''.
VERBOSE: Searching repository 'https://www.powershellgallery.com/api/v2/FindPackagesById()?id='Az.Advisor'' for ''.
VERBOSE: Searching repository 'https://www.powershellgallery.com/api/v2/FindPackagesById()?id='Az.Accounts'' for ''.
VERBOSE: Searching repository 'https://www.powershellgallery.com/api/v2/FindPackagesById()?id='Az.Aks'' for ''.
VERBOSE: Searching repository 'https://www.powershellgallery.com/api/v2/FindPackagesById()?id='Az.Accounts'' for ''.
VERBOSE: Searching repository 'https://www.powershellgallery.com/api/v2/FindPackagesById()?id='Az.AnalysisServices'' for ''.
VERBOSE: Searching repository 'https://www.powershellgallery.com/api/v2/FindPackagesById()?id='Az.Accounts'' for ''.
VERBOSE: Searching repository 'https://www.powershellgallery.com/api/v2/FindPackagesById()?id='Az.ApiManagement'' for ''.
VERBOSE: Searching repository 'https://www.powershellgallery.com/api/v2/FindPackagesById()?id='Az.Accounts'' for ''.
VERBOSE: Searching repository 'https://www.powershellgallery.com/api/v2/FindPackagesById()?id='Az.AppConfiguration'' for ''.
VERBOSE: Searching repository 'https://www.powershellgallery.com/api/v2/FindPackagesById()?id='Az.ApplicationInsights'' for ''.`

**Comment by timkatsapas** [comment]
I am just reiterating what remocoros stated above. 
Once its finished going through the "Searching Repository" it then only starts the install. Its definitely not optimized at all.

**Comment by dingmeng-xue** [comment]
I met this problem yesterday and PSGallery was very slow at that moment. To confirm the issue of PSGallery, you can try  `Install-Module -Name Microsoft.Graph`. Then report the issue to https://github.com/PowerShell/PowerShellGet/issues if it is hanging or slow as well.

I believe they will prioritize it when more issues are reported.

**Comment by dingmeng-xue** [comment]
@remcoros @CannonHax @timkatsapas , Thanks for your comments.

PSGallery is managed another team and performance is a known problem, my team cannot do more on this case. FYI, the bot will active this issue only when original author responds to it. The issue will be closed after 2 weeks.

**Comment by JLRishe** [comment]
I was experiencing the same issue so I tried installing with PowerShellGet 3.0.12. Not sure if that made it any faster, but at least I got a progress bar and could see (in verbose mode) that it was actually doing something other than making endless calls to `FindPackagesById()`. Finished in about 5 minutes.

```powershell
Install-psresource -Name Az -Scope CurrentUser -Repository PSGallery -Verbose
```

**Comment by BrunoJuchli** [comment]
This issue was raised on the PSResourceGet Repo https://github.com/PowerShell/PSResourceGet/issues/623

It was closed as duplicate of:
https://github.com/PowerShell/PSResourceGet/issues/69, https://github.com/PowerShell/PSResourceGet/issues/599 and https://github.com/PowerShell/PSResourceGet/issues/31

**Comment by masonzhang** [comment]
> This issue was raised on the PowerShell Get Repo: [PowerShell/PowerShellGet#623](https://github.com/PowerShell/PowerShellGet/issues/623)
> 
> It was closed as duplicate of: [PowerShell/PowerShellGet#69](https://github.com/PowerShell/PowerShellGet/issues/69), [PowerShell/PowerShellGet#599](https://github.com/PowerShell/PowerShellGet/issues/599) and [PowerShell/PowerShellGet#31](https://github.com/PowerShell/PowerShellGet/issues/31)

The issue has been there for 21 days, what's the progress?

I think it's a significant defect. It hangs for more than 30 minutes now.

**Comment by wontonotnow** [comment]
I guess this is bad timing on my part, I've never had to use Az before, >Install-Module Az (no parameters) shows no indication anything is happening....10 minutes so far, nada.  so here I am typing this---oh wait it's installing!

**Comment by masonzhang** [comment]
> I guess this is bad timing on my part, I've never had to use Az before, >Install-Module Az (no parameters) shows no indication anything is happening....10 minutes so far, nada. so here I am typing this---oh wait it's installing!

run the command before running Install-Module, you'll see what's going on:

`$debugPreference = "Continue"`

**Comment by dg-eparizzi** [comment]
Same thing, extremely slow.

Downloading at less than 1Mbps (100 Mbps connection)

![image](https://user-images.githubusercontent.com/72451869/164528005-2f509ec5-9e79-44fc-91d1-33b54be95c09.png)

**Comment by K-IronS** [comment]
Same here. Seeing multiple reports about the same issue. Why in the earth does this simple command have to search through all repositories with a ridiculous long timeout for each one?

To any admin or mod: Don't trust your monitoring tools. Do a manual test and try to install a module yourself, something is really wrong.

**Comment by dingmeng-xue** [comment]
Sorry for that. Our pipeline is blocked as well. AFAIK, PSGallery team is working on it but root cause is unknown yet.

**Comment by kilasuit** [comment]
> AFAIK, PSGallery team is working on it but root cause is unknown yet.

This issue is long standing & known due to the nature of how PowerShellGet (the module that Install-Module is a part of) deals with building the dependency tree of all the Az PowerShell Modules and cross dependencies of those modules. I have personally spent considerable amount of time in a number of different calls asking for a single module to be published that contains all the dependencies in a singlular [meta module structure akin to this one](url) that contains all required module as submodules instead of the current way, this however hasn't been seen as the way forward and hasn't been invested in, which I **still** think would make this so much simpler overall for all & would reduce network traffic as well as the install time needed for this module.

The [Az PowerShell team have however reccomended](https://twitter.com/azureposh/status/1518991534011547648?s=20&t=0T4bqIwXSNvAKwiGHYq_pQ) to use [Az.Tools.Installer module](https://www.powershellgallery.com/packages/Az.Tools.Installer) in the meantime (until PowerShellGet v3 is updated to do parallel dependency installs as per this [request I made](https://github.com/PowerShell/PowerShellGet/issues/69) however I personally would reccomend to use this [sample code](https://twitter.com/JustinWGrote/status/1172636154404818944?s=20&t=LrzPiIEJFReyWjfvhi-8PA) from @JustinGrote to install it instead of using Install-Module for the time being as this method is significantly quicker than the one 

cc @SteveL-MSFT @SydneyhSmith

**Comment by fidelcasto** [comment]
If you are installing at the computer scope, a way I found to make sure the background process is still running and to see where it's at is to look at the 'Date Modified' field of the folders for each module in C:\Program Files\WindowsPowerShell\Modules directory (or wherever your $env:PSModulePath is defined).

**Comment by roke-julian-lockhart** [comment]
@BrunoJuchli, all of the issues, that you have referenced within [`#issuecomment-1103841849`](https://github.com/Azure/azure-powershell/issues/17210#issuecomment-1103841849), are 404. Because this problem remains with, at the least, `Install-Module Microsoft.Graph.Beta -Scope AllUsers -Repository PSGallery -Force`, [^1] where is this tracked, now?

[^1]: [`PowerShell/PowerShell/discussions/27299#discussioncomment-16621853`](https://github.com/PowerShell/PowerShell/discussions/27299#discussioncomment-16621853)

**Comment by BrunoJuchli** [comment]
> [@BrunoJuchli](https://github.com/BrunoJuchli), all of the issues, that you have referenced within [`#issuecomment-1103841849`](https://github.com/Azure/azure-powershell/issues/17210#issuecomment-1103841849), are 404. Because this problem remains with, at the least, `Install-Module Microsoft.Graph.Beta -Scope AllUsers -Repository PSGallery -Force`, [1](#user-content-fn-1-4d30e0c7485f433b3eadbcec263b6dbc) where is this tracked, now?
> ## Footnotes
> 
>     1. [`PowerShell/PowerShell/discussions/27299#discussioncomment-16621853`](https://github.com/PowerShell/PowerShell/discussions/27299#discussioncomment-16621853) [↩](#user-content-fnref-1-4d30e0c7485f433b3eadbcec263b6dbc)

That's strange... I have no clue why they would delete the issues instead of closing them.
Two of the duplicate links still work, but to me it doesn't look like actual duplicates, a) they are PRs, b) they don't really look related.
AFAIR when I posted the comment I only really looked at https://github.com/PowerShell/PowerShellGet/issues/623, which corresponded to the problem in this issue. I didn't really look at all the issues it was marked duplicate of.

I might speculate that AI slop might have had a hand in this...

**Comment by kilasuit** [comment]
@RokeJulianLockhart `Install-Module` is part of the `technically deprecated` [PowerShellGet Module that is no longer under active development](https://github.com/powershell/powershellgetv2/) & is archived - note this is for PowerShellGet versions 2.x

There is a repo for [PowerShellGet 3.x](https://github.com/PowerShell/PowerShellGet/) that is a compatibility module to the supported and being improved [Microsoft.PowerShell.PSResourceGet](https://github.com/PowerShell/PSResourceGet/) module and in all honesty should be avoided if at all possible as in my opinion it's just more headaches than is needed.

Instead you should if using PowerShell 7.x check if you can replace `Install-Module` with `Install-PSResource` which comes from the [Microsoft.PowerShell.PSResourceGet Module](https://devblogs.microsoft.com/powershell/psresourceget-is-generally-available/) which comes with PowerShell 7.4.0 and above & superseeds the PowerShellGet Module.

If using PowerShell 5 I suggest an installation script that downloads and installs Microsoft.PowerShell.PSResourceGet that doesn't rely on using Install-module but instead downloads the nupkg and unzips that first.

> That's strange... I have no clue why they would delete the issues instead of closing them.
Two of the duplicate links still work, but to me it doesn't look like actual duplicates, a) they are PRs, b) they don't really look related.

As for some of the links @BrunoJuchli added like https://github.com/PowerShell/PowerShellGet/issues/623 - the PowerShell Team renamed the Repo to PSResourceGet so the correct links may actually be more likely to be like https://github.com/PowerShell/PSResourceGet/issues/623

**Comment by roke-julian-lockhart** [comment]
> As for some of the links [@BrunoJuchli](https://github.com/BrunoJuchli) added like https://github.com/PowerShell/PowerShellGet/issues/623 - the PowerShell Team renamed the Repo to PSResourceGet so the correct links may actually be more likely to be like [PowerShell/PSResourceGet#623](https://github.com/PowerShell/PSResourceGet/issues/623)

@kilasuit, considering [`discussions/22669#discussioncomment-3237667`](https://github.com/orgs/community/discussions/22669#discussioncomment-3237667), that's strange!
